### PR TITLE
Make `NewType` a class in 3.10

### PR DESCRIPTION
**Pull request by AlexWaygood**
`NewType` was changed in 3.10; it is now a class, rather than a function. https://bugs.python.org/issue44353

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Seems like this breaks pyright.

**Comment by dfoerster-athene** [comment]
I think the issue is here: https://github.com/python/typeshed/blob/8542b3518a36313af57b13227996168e592fe868/stubs/protobuf/google/protobuf/internal/enum_type_wrapper.pyi#L5. The TypeVar is bound to `int`, but is used instead with the distinct `_Syntax.V` NewType wrapper.

**Comment by AlexWaygood** [comment]
@erictraut, how tricky would it be to change the way pyright special-cases `NewType`? Would it be better for me to abandon this PR for now?

**Comment by erictraut** [comment]
@AlexWaygood, it's an easy change. I just implemented it, and it will be included in the next release of pyright (1.1.192). I can publish the new version later today (within the next 12 hours) to unblock your PR. You'd just need to update the CI scripts to point to pyright 1.1.192 as part of your PR.

**Comment by AlexWaygood** [comment]
> @AlexWaygood, it's an easy change. I just implemented it, and it will be included in the next release of pyright (1.1.192). I can publish the new version later today (within the next 12 hours) to unblock your PR. You'd just need to update the CI scripts to point to pyright 1.1.192 as part of your PR.

@erictraut fantastic, thank you!

**Comment by erictraut** [comment]
I published pyright 1.1.192, which contains this change. You should be able to update `workflows/tests.yml` and `tests/pyright_test.py` to point to the new version of pyright. Search for "1.1.187", which is the version that's currently used in typeshed.

**Comment by AlexWaygood** [comment]
> I published pyright 1.1.192, which contains this change. You should be able to update `workflows/tests.yml` and `tests/pyright_test.py` to point to the new version of pyright. Search for "1.1.187", which is the version that's currently used in typeshed.

The pyright tests are now green 🥳🥳

**Review by Akuli — APPROVED** [review]
I'm a bit surprised that stubtest didn't warn about this. Maybe it is because stubtest is based on mypy, and mypy special-cases this?

**Comment by AlexWaygood** [comment]
> I'm a bit surprised that stubtest didn't warn about this. Maybe it is because stubtest is based on mypy, and mypy special-cases this (I don't know if it does)?

Yeah it's a weird one. My prior with the `typing` module is that most of it is special-cased by type-checkers, so the standard tests may therefore be of limited use.

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -83,6 +83,13 @@ if sys.version_info >= (3, 10):
     Concatenate: _SpecialForm = ...
     TypeAlias: _SpecialForm = ...
     TypeGuard: _SpecialForm = ...
+    class NewType:
+        def __init__(self, name: str, tp: type) -> None: ...
+        def __call__(self, x: _T) -> _T: ...
+        __supertype__: type
```
Comment (srittau) [inline stdlib/typing.pyi:89]: This also has `__or__` and `__ror__` methods which we should include.

**Review by srittau — CHANGES_REQUESTED** [review]

**Comment by hauntsaninja** [comment]
It's allowlisted: https://github.com/python/typeshed/blob/1fa1270e00bd6ce351cf3a02ed785df9a8c17176/tests/stubtest_allowlists/py3_common.txt#L246

Fwiw, I'm -0 on this change (aka I don't really mind it, but I won't hit "merge" myself).

This wasn't obvious to me when I first started contributing to typeshed, but I've come to conclude that trying to stub things in typing.py accurately is just fraught (not necessarily relevant to this PR):
1) type checkers special case the hell out of most of these defs, so it doesn't accomplish much
1a) often changes just cause churn for type checkers whose special casing is now broken, as above
2) typing.py internals have historically been extremely prone to change
3) there have been cases where changing things in typing.py can cause worse errors for users
3a) more accurate stubs can cause confusion for some enterprising users who use stub definitions to conclude things about type checker behaviour for typing.py constructs
4) analysing typing.pyi is on the critical path for any type checking operation, so keeping things simple is nice

**Comment by hauntsaninja** [comment]
Oh actually I take it back. The addition of `__supertype__` here seems useful.

**Comment by AlexWaygood** [comment]
> Fwiw, I'm -0 on this change (aka I don't really mind it, but I won't hit "merge" myself).

Yeah, I get all that. I still think this change would be nice, as a relatively pain-free way to get the stubs closer to their runtime implementation.

**Comment by AlexWaygood** [comment]
I am Unsure whether `__or__` should be typed like this  (does this even make sense???);

```python
def __or__(self: Self, other: Type[_T) -> Type[Union[Self, Type[_T]]]: ...
```

Or, more simply, like this:

```python
def __or__(self, other: type) ->_SpecialForm: ...
```

...Thoughts?

**Comment by srittau** [comment]
> This wasn't obvious to me when I first started contributing to typeshed, but I've come to conclude that trying to stub things in typing.py accurately is just fraught (not necessarily relevant to this PR):
> 
> 1. type checkers special case the hell out of most of these defs, so it doesn't accomplish much
>    1a) often changes just cause churn for type checkers whose special casing is now broken, as above

I see the primary benefit of accurate types in `typing.pyi` not for type checkers, but other tools, like code completion or other, unknown uses. So all things being equal I prefer accurate types over non-accurate types in `typing.pyi`.

**Comment by hauntsaninja** [comment]
Actually, it might be nice to make this unconditionally a class, since older Pythons have the `__supertype__` attribute as well.

**Comment by AlexWaygood** [comment]
> Oh actually I take it back. The addition of `__supertype__` here seems maybe useful.

The `__or__` and `__ror__` methods are also important, as otherwise `NewType` can't be used in PEP 604 union-type expressions. This was the original motivation for changing the implementation of `NewType` in 3.10 (making it kind of hilarious that I left them out of the first draft of the PR).

**Comment by hauntsaninja** [comment]
At least for mypy, or / ror get special cased. E.g., PEP 604 works with future annotations before 3.10 despite no typeshed or / ror.

**Comment by AlexWaygood** [comment]
> At least for mypy, or / ror get special cased.

Not so for PyCharm, if memory serves correctly (though that might have changed since I last used it)

File: stdlib/typing.pyi [hunk 2]
```diff
@@ -83,6 +83,15 @@ if sys.version_info >= (3, 10):
     Concatenate: _SpecialForm = ...
     TypeAlias: _SpecialForm = ...
     TypeGuard: _SpecialForm = ...
+    class NewType:
+        def __init__(self, name: str, tp: type) -> None: ...
+        def __call__(self, x: _T) -> _T: ...
+        def __or__(self, other: Any) -> _SpecialForm: ...
+        def __or__(self, other: Any) -> _SpecialForm: ...
```
Comment (hauntsaninja) [inline stdlib/typing.pyi:90]: ror :-)

Comment (AlexWaygood) [reply stdlib/typing.pyi:90]: Dammit

**Comment by AlexWaygood** [comment]
> I am Unsure whether `__or__` should be typed like this (does this even make sense???):
> 
> ```python
> def __or__(self: Self, other: Type[_T]) -> Type[Union[Self, Type[_T]]]: ...
> ```
> 
> Or, more simply, like this:
> 
> ```python
> def __or__(self, other: type) -> _SpecialForm: ...
> ```
> 
> ...Thoughts?

I went for the simpler option, on the assumption that it's all special-cased to some extent by type-checkers anyway, and because it's more similar to `type.__or__`. I changed the parameter type to `Any`, though, as I realised the other object doesn't _have_ to be a type — it could be another `NewType`, or a `TypeVar`, or quite a few things really.

**Review by srittau — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks for your help, all! 🎉

**Pull request merged**